src/lib.rs

```diff
@@ -123,15 +122,25 @@ unsafe extern "C" fn p_init(
             return PSA_ERROR_GENERIC_ERROR;
         }
     };
-    let tpm_provider_uuid = Uuid::parse_str("1e4954a4-ff21-46d3-ab0c-661eeb667e1d").unwrap();
     let provider_id = match providers
         .iter()
-        .filter(|p| p.uuid == tpm_provider_uuid)
-        .last()
+        .filter(|p| {
+            if cfg!(feature = "tpm") {
+                // Only keep the TPM provider.
+                p.uuid == Uuid::parse_str("1e4954a4-ff21-46d3-ab0c-661eeb667e1d").unwrap()
+            } else if cfg!(feature = "pkcs11") {
+                // Only keep the PKCS11 provider.
+                p.uuid == Uuid::parse_str("30e39502-eba6-4d60-a4af-c518b7f5e38f").unwrap()
+            } else {
+                // Filter the Core provider.
+                p.uuid != Uuid::parse_str("47049873-2a43-4845-9d72-831eab668784").unwrap()
+            }
```

Comment:
You should either document this or remove the `else` bit from the 2nd `if` - if both `tpm` and `pkcs11` are enabled, `tpm` takes precedence.